src/python/pants/backend/common/targets/app.py

```diff
@@ -0,0 +1,259 @@
+
+    real_rel_path = rel_path or self._parse_context.rel_path
+
+    if relative_to:
+      base = os.path.join(get_buildroot(), real_rel_path, relative_to)
+      mapper = RelativeToMapper(base)
+    else:
+      mapper = mapper or RelativeToMapper(os.path.join(get_buildroot(), real_rel_path))
+
+    return BundleProps(real_rel_path, mapper, fileset)
+
+  def create_bundle_props(self, bundle):
+    rel_path = getattr(bundle, 'rel_path', None)
+    mapper = getattr(bundle, 'mapper', None)
+    relative_to = getattr(bundle, 'relative_to', None)
+    fileset = getattr(bundle, 'fileset', None)
+    return self(rel_path, mapper, relative_to, fileset)
+
+
+class BundleField(tuple, PayloadField):
+  """A tuple subclass that mixes in PayloadField.
+
+  Must be initialized with an iterable of Bundle instances.
+  """
+
+  @staticmethod
+  def _hash_bundle(bundle):
+    hasher = sha1()
+    hasher.update(bundle.rel_path)
+    for abs_path in sorted(bundle.filemap.keys()):
+      buildroot_relative_path = os.path.relpath(abs_path, get_buildroot())
+      hasher.update(buildroot_relative_path)
+      hasher.update(bundle.filemap[abs_path])
+      if os.path.isfile(abs_path):
+        # Update with any additional string to differentiate empty file with non-existing file.
+        hasher.update('e')
+        with open(abs_path, 'rb') as f:
+          hasher.update(f.read())
+    return hasher.hexdigest()
+
+  def _compute_fingerprint(self):
+    return combine_hashes(map(BundleField._hash_bundle, self))
+
+
+class App(Target):
+  """A deployable application.
+
+  Invoking the ``bundle`` goal on one of these targets creates a
+  self-contained artifact suitable for deployment on some other machine.
+  The artifact contains the executable, its dependencies, and
+  extra files like config files, startup scripts, etc.
+
+  :API: public
+  """
+  class InvalidArchiveType(Exception):
+    """Raised when archive type defined in Target is invalid"""
+
+  def __init__(self, name=None, payload=None, **kwargs):
+    if payload.archive and payload.archive not in Archive.TYPE_NAMES:
+      raise self.InvalidArchiveType(
+        'Given archive type "{}" is invalid, choose from {}.'.format(
+          payload.archive, list(Archive.TYPE_NAMES)))
+    super(App, self).__init__(name=name, payload=payload, **kwargs)
+
+  def globs_relative_to_buildroot(self):
+    buildroot = get_buildroot()
+    globs = []
+    for bundle in self.bundles:
+      fileset = bundle.fileset
+      if fileset is None:
+        continue
+      elif hasattr(fileset, 'filespec'):
+        globs += bundle.fileset.filespec['globs']
+      else:
+        # NB(nh): filemap is an OrderedDict, so this ordering is stable.
+        globs += [fast_relpath(f, buildroot) for f in bundle.filemap.keys()]
+    super_globs = super(App, self).globs_relative_to_buildroot()
+    if super_globs:
+      globs += super_globs['globs']
+    return {'globs': globs}
+
+  @classmethod
+  def _binary_target_type(cls):
```

Comment:
This should probably be called `binary_target_type`, as the leading underscore strongly implies "you can change this without checking for uses outside this file", but subclasses must implement this, so that implication would be wrong in this case.